In [13]:
import numpy as np
import pandas as pd

from sklearn.linear_model import SGDClassifier
from sklearn.svm import SVC, LinearSVC
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score, roc_auc_score
from sklearn.model_selection import cross_val_score, GridSearchCV, ShuffleSplit
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.feature_selection import SelectFromModel

# Import Data

In [14]:
data = pd.read_csv('./data/train.csv')
data.drop('id', axis = 1, inplace = True)
X_train = np.asarray(data[data.columns[range(1, data.shape[1])]], dtype = np.double)
y_train = np.asarray(data[['label']], dtype = np.double).ravel()
data.head()

label  feat1  feat2  feat3  feat4  feat5  feat6  feat7  feat8  feat9  \
0      1      7      0      3      0      2      3      0      6      0   
1      1      0     11      0      0     10      1      0      0      4   
2      0      9      0      3      0      1      3      0      4      0   
3      0      0      9      3      2     25      0      4      0      0   
4      0      0      0      0      0      2      5      0      0      0   

    ...     feat207  feat208  feat209  feat210  feat211  feat212  feat213  \
0   ...           3        4        2        2        0       13        0   
1   ...           0        2        0        0        2        8        1   
2   ...          48       11        2        0        0        4        0   
3   ...           1       14        1        0        0        0        3   
4   ...           3       12        0        3        0        4        0   

   feat214  feat215  feat216  
0       11        1        3  
1       13        0        4  
2        2        0        0  
3        0       17        1  
4       24        4        0  

[5 rows x 217 columns]

In [15]:
data.groupby(['label']).count()

feat1  feat2  feat3  feat4  feat5  feat6  feat7  feat8  feat9  feat10  \
label                                                                          
0        327    327    327    327    327    327    327    327    327     327   
1        573    573    573    573    573    573    573    573    573     573   

        ...     feat207  feat208  feat209  feat210  feat211  feat212  feat213  \
label   ...                                                                     
0       ...         327      327      327      327      327      327      327   
1       ...         573      573      573      573      573      573      573   

       feat214  feat215  feat216  
label                             
0          327      327      327  
1          573      573      573  

[2 rows x 216 columns]

# Data preprocessing

In [16]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_train

array([[ 1.67857588, -0.6426044 ,  0.27134247, ..., -0.09405589,
        -0.57018666, -0.29529026],
       [-0.61884494,  1.01109551, -0.70941344, ...,  0.04586196,
        -0.76376191, -0.16160755],
       [ 2.33498182, -0.6426044 ,  0.27134247, ..., -0.7236862 ,
        -0.76376191, -0.6963384 ],
       ...,
       [-0.61884494, -0.6426044 , -0.70941344, ...,  0.32569765,
        -0.76376191,  1.44258501],
       [-0.61884494, -0.34193169, -0.05557617, ...,  0.11582088,
        -0.57018666,  0.10575788],
       [ 1.67857588, -0.6426044 , -0.3824948 , ...,  0.95532796,
        -0.76376191, -0.29529026]])

In [17]:
# Convert feature matrix into DataFrame
df = pd.DataFrame(X_train)

# View the data frame
df

0         1         2         3         4         5         6    \
0    1.678576 -0.642604  0.271342 -0.670364 -0.255059 -0.183106 -0.517856   
1   -0.618845  1.011096 -0.709413 -0.670364  1.180770 -0.566352 -0.517856   
2    2.334982 -0.642604  0.271342 -0.670364 -0.434538 -0.183106 -0.517856   
3   -0.618845  0.710423  0.271342  1.229881  3.872949 -0.757975  1.730976   
4   -0.618845 -0.642604 -0.709413 -0.670364 -0.255059  0.200140 -0.517856   
5    0.037561  1.161432 -0.709413  0.279758  1.180770 -0.374729 -0.517856   
6   -0.618845 -0.492268 -0.709413  0.279758 -0.434538  0.966632  0.044352   
7   -0.290642 -0.191595 -0.709413 -0.670364 -0.434538 -0.374729 -0.517856   
8    3.647794 -0.642604 -0.709413  2.180004 -0.614016 -0.374729  0.606560   
9    0.365764 -0.041259  1.252098 -0.670364  0.283377 -0.183106  2.293184   
10  -0.618845 -0.341932 -0.382495 -0.670364  0.462855  0.966632  0.044352   
11   2.006779 -0.492268 -0.382495 -0.670364 -0.614016 -0.566352  0.044352   
12  -0.618845 -0.642604 -0.709413 -0.670364 -0.614016 -0.757975 -0.517856   
13   0.693967 -0.492268 -0.055576 -0.670364 -0.614016  0.008517 -0.517856   
14  -0.618845 -0.492268  0.271342 -0.670364 -0.614016 -0.757975 -0.517856   
15  -0.618845 -0.492268  2.559773 -0.670364 -0.614016 -0.374729  1.168768   
16   3.319591 -0.492268 -0.382495  0.279758 -0.614016 -0.566352  0.044352   
17   0.037561 -0.642604 -0.709413 -0.670364 -0.614016  0.583386 -0.517856   
18   0.365764 -0.191595  0.271342 -0.670364 -0.434538  0.391763  0.606560   
19   1.350373 -0.642604 -0.055576 -0.670364 -0.614016 -0.757975  0.044352   
20  -0.618845 -0.642604  0.271342 -0.670364 -0.255059  2.691239  0.044352   
21  -0.290642  1.011096 -0.382495  0.279758  0.283377 -0.183106 -0.517856   
22  -0.618845 -0.642604 -0.709413  0.279758 -0.614016  0.200140 -0.517856   
23  -0.290642  2.664795 -0.055576  1.229881 -0.255059  0.008517 -0.517856   
24  -0.618845  0.109077 -0.382495 -0.670364 -0.614016 -0.757975 -0.517856   
25  -0.618845 -0.041259 -0.382495 -0.670364  6.026693  0.008517 -0.517856   
26  -0.618845 -0.642604 -0.382495 -0.670364 -0.614016  1.924747 -0.517856   
27   0.037561 -0.642604 -0.709413 -0.670364 -0.614016 -0.374729 -0.517856   
28  -0.290642  0.259414 -0.055576 -0.670364 -0.255059  0.008517 -0.517856   
29   1.350373 -0.642604 -0.055576 -0.670364 -0.434538 -0.374729 -0.517856   
..        ...       ...       ...       ...       ...       ...       ...   
870 -0.618845 -0.492268 -0.382495 -0.670364 -0.614016 -0.757975 -0.517856   
871 -0.618845 -0.492268  1.579017  1.229881 -0.614016  0.391763 -0.517856   
872 -0.618845  1.913114 -0.709413 -0.670364 -0.255059 -0.757975 -0.517856   
873 -0.618845  0.560086 -0.055576  0.279758 -0.434538 -0.757975 -0.517856   
874  0.037561 -0.341932 -0.382495  2.180004 -0.434538  2.691239 -0.517856   
875 -0.618845  2.063450 -0.709413 -0.670364 -0.434538 -0.757975 -0.517856   
876 -0.618845 -0.642604 -0.382495  1.229881 -0.075580  1.158255  1.168768   
877 -0.618845 -0.341932 -0.055576  1.229881 -0.614016 -0.566352 -0.517856   
878 -0.618845  1.612441 -0.709413 -0.670364  2.796078 -0.374729 -0.517856   
879 -0.618845 -0.341932  0.271342  4.080249 -0.075580  0.966632 -0.517856   
880  1.022170 -0.642604 -0.055576 -0.670364 -0.614016 -0.566352 -0.517856   
881 -0.618845 -0.341932 -0.055576 -0.670364 -0.255059 -0.183106 -0.517856   
882 -0.290642 -0.191595 -0.709413 -0.670364 -0.075580 -0.757975 -0.517856   
883 -0.618845 -0.191595  0.598261  0.279758  1.001291 -0.566352  0.606560   
884 -0.290642  0.409750 -0.382495 -0.670364 -0.434538 -0.757975  0.044352   
885 -0.618845  0.109077  0.598261  0.279758  2.078163  3.074485 -0.517856   
886 -0.290642 -0.642604 -0.709413  5.030372 -0.434538 -0.757975 -0.517856   
887  0.365764 -0.041259 -0.709413  0.279758 -0.255059  0.966632 -0.517856   
888 -0.618845 -0.492268 -0.709413 -0.670364 -0.075580 -0.757975 -0.517856   
889 -0.618845  1.311768  0.598261  2.180004  1.001291  0.391763  0.044352   
890 -0

In [18]:
# Create correlation matrix
corr_matrix = df.corr().abs()

# Select upper triangle of correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

# Find index of feature columns with correlation greater than 0.95
to_drop = [column for column in upper.columns if any(upper[column] > 0.8)]

In [19]:
# Drop features 
df.drop(df.columns[to_drop], axis=1, inplace=True)

In [20]:
X_train = df.as_matrix()
print(X_train.shape)

(900, 204)


# Testing different models

### SGDClassifier

In [ ]:
best_params = {
    'loss': 'squared_hinge', 
    'penalty': 'elasticnet',
    'alpha': 1e-4,
}

In [ ]:
clf = SGDClassifier(max_iter=1000, random_state=0, **best_params)

In [ ]:
params = {
#     'loss': ['hinge', 'log', 'modified_huber', 'squared_hinge', 'perceptron', 
#              'squared_loss', 'huber', 'epsilon_insensitive', 'squared_epsilon_insensitive'],
#     'penalty': ['l2', 'l1', 'elasticnet'],
    'alpha': [1e-5, 1e-4, 1e-3, 1e-2],
}

In [ ]:
grid_search = GridSearchCV(clf, 
                           params, 
                           scoring='roc_auc', 
                           n_jobs=-2, 
                           cv=5, 
                           verbose=1, 
                           return_train_score=False)
grid_search.fit(X_train, y_train)

In [ ]:
print(grid_search.best_score_)
print(grid_search.best_params_)
grid_search.best_estimator_

In [ ]:
pd.DataFrame(grid_search.cv_results_)

*Лучший классификатор:* ~0.933, 0.90

SGDClassifier(alpha=0.0001, average=False, class_weight=None, epsilon=0.1,
       eta0=0.0, fit_intercept=True, l1_ratio=0.15,
       learning_rate='optimal', loss='perceptron', max_iter=5000,
       n_iter=None, n_jobs=1, penalty='elasticnet', power_t=0.5,
       random_state=0, shuffle=True, tol=None, verbose=0, warm_start=False)

### SVC

In [115]:
best_params = {
    'kernel': 'rbf', 
    'C': 0.0001, 
    'gamma': 1,
}

In [116]:
clf = SVC(random_state=0, **best_params)

In [117]:
params = {
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'gamma': np.geomspace(1e-3, 1e3, 10),
    'C': np.geomspace(1e-3, 1e3, 10),
#     'gamma': np.linspace(0.001, 2, 30),
#     'class_weight': [None, 'balanced'],
}

In [118]:
grid_search = GridSearchCV(clf, 
                           params, 
                           scoring='roc_auc', 
                           n_jobs=-2, 
                           cv=5, 
                           verbose=1, 
                           return_train_score=False)
grid_search.fit(X_train, y_train)

Fitting 5 folds for each of 400 candidates, totalling 2000 fits


[Parallel(n_jobs=-2)]: Done  58 tasks      | elapsed:    1.7s
[Parallel(n_jobs=-2)]: Done 358 tasks      | elapsed:   10.1s
[Parallel(n_jobs=-2)]: Done 858 tasks      | elapsed:   24.1s
[Parallel(n_jobs=-2)]: Done 1558 tasks      | elapsed:   43.0s
[Parallel(n_jobs=-2)]: Done 2000 out of 2000 | elapsed:   55.3s finished


GridSearchCV(cv=5, error_score='raise',
       estimator=SVC(C=0.0001, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=1, kernel='rbf',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=-2,
       param_grid={'kernel': ['linear', 'poly', 'rbf', 'sigmoid'], 'gamma': array([1.00000e-03, 4.64159e-03, 2.15443e-02, 1.00000e-01, 4.64159e-01,
       2.15443e+00, 1.00000e+01, 4.64159e+01, 2.15443e+02, 1.00000e+03]), 'C': array([1.00000e-03, 4.64159e-03, 2.15443e-02, 1.00000e-01, 4.64159e-01,
       2.15443e+00, 1.00000e+01, 4.64159e+01, 2.15443e+02, 1.00000e+03])},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
       scoring='roc_auc', verbose=1)

In [119]:
print(grid_search.best_score_)
print(grid_search.best_params_)
grid_search.best_estimator_

0.9814463896134378
{'C': 10.0, 'gamma': 0.021544346900318832, 'kernel': 'rbf'}


SVC(C=10.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.021544346900318832,
  kernel='rbf', max_iter=-1, probability=False, random_state=0,
  shrinking=True, tol=0.001, verbose=False)

In [120]:
pd.DataFrame(grid_search.cv_results_)

mean_fit_time  mean_score_time  mean_test_score param_C param_gamma  \
0         0.089298         0.017509         0.905703   0.001       0.001   
1         0.101155         0.023144         0.914828   0.001       0.001   
2         0.117436         0.026064         0.908518   0.001       0.001   
3         0.137039         0.034278         0.887522   0.001       0.001   
4         0.082640         0.020739         0.905703   0.001  0.00464159   
5         0.164444         0.037537         0.930901   0.001  0.00464159   
6         0.184971         0.039315         0.921717   0.001  0.00464159   
7         0.175840         0.039910         0.878900   0.001  0.00464159   
8         0.118858         0.026546         0.905703   0.001   0.0215443   
9         0.171133         0.037244         0.942895   0.001   0.0215443   
10        0.197060         0.044075         0.974626   0.001   0.0215443   
11        0.189293         0.041691         0.870701   0.001   0.0215443   
12        0.112617         0.024810         0.905703   0.001         0.1   
13        0.161615         0.027327         0.971524   0.001         0.1   
14        0.175168         0.043565         0.961025   0.001         0.1   
15        0.185326         0.040737         0.863242   0.001         0.1   
16        0.120013         0.026359         0.905703   0.001    0.464159   
17        0.155708         0.025132         0.970306   0.001    0.464159   
18        0.191042         0.046063         0.903014   0.001    0.464159   
19        0.179553         0.041503         0.857848   0.001    0.464159   
20        0.112427         0.025056         0.905703   0.001     2.15443   
21        0.159695         0.026112         0.970306   0.001     2.15443   
22        0.198012         0.046988         0.896839   0.001     2.15443   
23        0.166414         0.038193         0.858322   0.001     2.15443   
24        0.113304         0.025077         0.905703   0.001          10   
25        0.163750         0.026648         0.970306   0.001          10   
26        0.177434         0.042239         0.896839   0.001          10   
27        0.163627         0.037027         0.857495   0.001          10   
28        0.116243         0.026037         0.905703   0.001     46.4159   
29        0.159856         0.026308         0.970306   0.001     46.4159   
..             ...              ...              ...     ...         ...   
370       0.174822         0.037000         0.981446    1000   0.0215443   
371       0.100093         0.018985         0.726402    1000   0.0215443   
372       0.271716         0.014093         0.916189    1000         0.1   
373       0.174709         0.027087         0.970306    1000         0.1   
374       0.186652         0.038942         0.965400    1000         0.1   
375       0.115255         0.021152         0.698294    1000         0.1   
376       0.250016         0.013297         0.916189    1000    0.464159   
377       0.180821         0.029754         0.970306    1000    0.464159   
378       0.204371         0.045033         0.910821    1000    0.464159   
379       0.122028         0.022029         0.700048    1000    0.464159   
380       0.259820         0.013653         0.916189    1000     2.15443   
381       0.154514         0.025389         0.970306    1000     2.15443   
382       0.184680         0.041558         0.896839    1000     2.15443   
383       0.095808         0.018811         0.708120    1000     2.15443   
384       0.265021         0.013480         0.916189    1000          10   
385       0.161518         0.026503         0.970306    1000          10   
386       0.164689         0.037076         0.896839    1000          10   
387       0.089700         0.018817         0.708309    1000          10   
388       0.238512         0.012765         0.916189    1000     46.4159   
389       0.163110         0.026621         0.970306    1000     46.4159   
390       0.168423         0.038105         0.896839 

*Лучший классификатор*: ~0.972, 

SVC(C=0.0001, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=1, kernel='poly',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False)

### LinearSVC

In [ ]:
best_params = { 
    'dual': False,
    'loss': 'squared_hinge',
    'penalty': 'l1',
    'C': 0.06,
}

In [ ]:
clf = LinearSVC(**best_params, random_state=0)

In [ ]:
params = {
#     'penalty': ['l1', 'l2'],
#     'loss': ['hinge', 'squared_hinge'],
#     'dual': [True, False],
    'C': [1e-4, 1e-3, 1e-2, 1e-1, 1, 10, 100]
}

In [ ]:
grid_search = GridSearchCV(clf, 
                           params, 
                           scoring='roc_auc', 
                           n_jobs=-2, 
                           cv=5, 
                           verbose=1, 
                           return_train_score=False)
grid_search.fit(X_train, y_train)

In [ ]:
print(grid_search.best_score_)
print(grid_search.best_params_)
grid_search.best_estimator_

In [ ]:
pd.DataFrame(grid_search.cv_results_)

Лучший классификатор: ~0.926,

LinearSVC(C=0.1, class_weight=None, dual=False, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l1', random_state=0, tol=0.0001,
     verbose=0)

### LogisticRegression

In [ ]:
best_params = {
    'penalty': 'l2', 
    'C': 0.2,
    'max_iter': 500,
}

In [ ]:
clf = LogisticRegression(**best_params, random_state=0)

In [ ]:
params = {
#     'penalty': ['l1', 'l2'],
    'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'],
#     'dual': [True, False],
    'C': [1e-4, 1e-3, 1e-2, 1e-1, 1, 10, 100]
}

In [ ]:
grid_search = GridSearchCV(clf, 
                           params, 
                           scoring='roc_auc', 
                           n_jobs=-2, 
                           cv=5, 
                           verbose=1, 
                           return_train_score=False)
grid_search.fit(X_train, y_train)

In [ ]:
print(grid_search.best_score_)
print(grid_search.best_params_)
grid_search.best_estimator_

In [ ]:
pd.DataFrame(grid_search.cv_results_)

Лучший классификатор: ~0.930,

LogisticRegression(C=0.1, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=500, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=0, solver='saga', tol=0.0001,
          verbose=0, warm_start=False)

# Fit final classifier

In [ ]:
estimator = SVC(C=0.0001, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=1, kernel='poly',
  max_iter=-1, probability=True, random_state=0, shrinking=True,
  tol=0.001, verbose=False)
estimator.fit(X_train, y_train)

# Test data estimation

In [ ]:
test = pd.read_csv('./data/test.csv')
X_test = np.asarray(test[test.columns[range(1, test.shape[1])]])
X_test = scaler.transform(X_test)

In [ ]:
X_test

In [ ]:
# test_predict = estimator.predict(X_test)
test_predict = estimator.predict_proba(X_test)[:, 1]
print(test_predict)

In [ ]:
test['label'] = test_predict
test[['id', 'label']].to_csv('submit2.csv', sep=',', index=False)